### PR TITLE
CLOUDCI-1286: Add build timeout to pipeline run context

### PR DESCRIPTION
**Pull request by pavanmn22**
### Description

&lt;Add detailed description for reviewers here.&gt;

### Submitter checklist

- [ ] Change has been tested (on a back-end cluster)
- [ ] (If applicable) Jira backlog item ID added to the PR title and the [changelog.yaml] entry
- [ ] [changelog.yaml] entry with upgrade notes is prepared and appropriate for the audience affected by the change (users or developer, depending on the change)
- [ ] Semantic version diffed against [last release][releases] and updated accordingly. In this project the version has to be maintained here:
    - [/charts/steward/Chart.yaml](https://github.com/SAP/stewardci-core/blob/master/charts/steward/Chart.yaml) (`version` and `appVersion`)

In case dependencies have been updated:
- [ ] Links to external changelogs, since the last release of our component, added to the [changelog.yaml] entry (description).
- [ ] Changelogs read thoroughly, potential impact described, upgrade notes prepared (if necessary)
- [ ] Check if dependency updates affect our semantic version increment.

### Reviewer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There is at least 1 approval for the pull request and no outstanding requests for change
- [ ] All voter checks have passed
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] The Pull Request title is understandable and reflects the changes well
- [ ] The Pull Request description is understandable and well documented
- [ ] [changelog.yaml] entry for this Pull Request has been added
    - [ ] Changelog entry contains all required information
    - [ ] 'Upgrade notes' are documented in changelog.yaml (if required)

[changelog.yaml]: https://github.com/SAP/stewardci-core/changelog.yaml
[releases]: https://github.com/SAP/stewardci-core/releases

### CODE REVIEW

File: charts/steward/crds/pipelineruns.yaml [hunk 1]
```diff
@@ -78,6 +78,9 @@ spec:
                 - run
                 - abort
                 default: run
+              "timeout": ###
+                type: string
+                pattern: '^[0-9]*m?$'
```
Comment (anbrsap) [inline charts/steward/crds/pipelineruns.yaml:83]: We shouldn't introduce another specification different from that of [`time.ParseDuration()`](https://pkg.go.dev/time#ParseDuration). An exception might be that negative durations are not allowed.

File: docs/examples/pipelinerun_timeout.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+apiVersion: steward.sap.com/v1alpha1
+kind: PipelineRun
+metadata:
+  generateName: sleep
+spec:
+  timeout: "1d"
```
Comment (anbrsap) [inline docs/examples/pipelinerun_timeout.yaml:6]: `d` is not allowed by [`time.ParseDuration()`](https://pkg.go.dev/time#ParseDuration).

File: pkg/runctl/run_manager.go [hunk 3]
```diff
@@ -480,9 +477,14 @@ func (c *runManager) createTektonTaskRun(ctx context.Context, runCtx *runContext
 	namespace := runCtx.runNamespace
 	serviceAccountSecretName, err := c.getServiceAccountSecretName(ctx, runCtx)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	log.Printf("Pipeline Run: %+v", runCtx.pipelineRun.GetSpec())
```
Comment (anbrsap) [inline pkg/runctl/run_manager.go:482]: Use klog instead.

File: pkg/runctl/run_manager.go [hunk 4]
```diff
@@ -480,9 +477,14 @@ func (c *runManager) createTektonTaskRun(ctx context.Context, runCtx *runContext
...
+	timeout := runCtx.pipelineRunsConfig.Timeout
+	pipelineRunTimeout := runCtx.pipelineRun.GetSpec().Timeout
+	if pipelineRunTimeout != nil {
+		timeout = pipelineRunTimeout
 	}
```
Comment (anbrsap) [inline pkg/runctl/run_manager.go:487]: You may encapsulate this selection logic into a function to separate it from the main code that just assembles the object:

```suggestion
	timeout := func() metav1.Duration {
		timeout := runCtx.pipelineRun.GetSpec().Timeout
		if timeout == nil {
			timeout = &runCtx.pipelineRunsConfig.Timeout
		}
		return *timeout
	}
```

File: docs/examples/pipelinerun_timeout.yaml (hunk 2 above)
Comment (rinckm) [reply docs/examples/pipelinerun_timeout.yaml:6]: This was committed by mistake. It was a check if CRD pattern works. Will be fixed.

File: pkg/runctl/run_manager.go (hunk 3 above)
Comment (rinckm) [reply pkg/runctl/run_manager.go:482]: Will be removed. Debugging only.

File: charts/steward/crds/pipelineruns.yaml (hunk 1 above)
Comment (pavanmn22) [reply charts/steward/crds/pipelineruns.yaml]: Modified the regex to: ([0-9]+h)?([0-9]+m)?([0-9]+s)?, this allows setting the timeout in hrs, min, and secs

File: pkg/runctl/run_manager.go (hunk 3 above)
Comment (pavanmn22) [reply pkg/runctl/run_manager.go]: Removed the log statement.

File: pkg/runctl/run_manager.go (hunk 4 above)
Comment (pavanmn22) [reply pkg/runctl/run_manager.go]: Moved fetching of timeout into a separate function.

File: docs/backend-api/README.md [hunk 5]
```diff
@@ -202,6 +202,7 @@ A simple PipelineRun resource example can be found in [docs/examples/pipelinerun
 | `spec.logging` | (object,optional) The logging configuration. |
 | `spec.logging.elasticsearch` | (object,optional) The configuration for pipeline logging to Elasticsearch. If not specified, logging to Elasticsearch is disabled and the default Jenkins log implementation is used (stdout of Jenkinsfile Runner container). |
 | `spec.logging.elasticsearch.runID` | (any,optional) The JSON value that should be set as field `runId` in each log entry in Elasticsearch. It can be any JSON value (`null`, boolean, number, string, list, map). |
+| `spec.timeout` | (string,optional) The timeout value specified for a steward pipeline run in duration string format of composed of whole numbers, each with a unit suffix, such as "300m", "15h" or "2h45m". Valid time units are "s", "m" and "h". |
```
Comment (rinckm) [inline docs/backend-api/README.md:205]:
```suggestion
| `spec.timeout` | (string,optional) The timeout value specified for a steward pipeline run. The duration string format is composed of whole numbers, each with a unit suffix, such as "300m", "15h" or "2h45m". Valid time units are "s", "m" and "h". |
```

File: pkg/runctl/run_manager_test.go [hunk 6]
```diff
@@ -2054,3 +2054,54 @@ func (*testHelper1) prepareMocksWithSpec(ctrl *gomock.Controller, spec *stewardv
...
+
+	config:= &cfg.PipelineRunsConfigStruct{
+	  Timeout: defaultTimeout,
+	}
+	runCtx:= &runContext{
+  			pipelineRun:        k8sPipelineRun,
+  			pipelineRunsConfig: config,
+  }
+
+  //exercise
```
Comment (rinckm) [inline pkg/runctl/run_manager_test.go:2082]: Please use all off the following markers (uppercase)

// SETUP
// EXERCISE
// VERIFY

**Review by rinckm — APPROVED** [review]

**Pull request merged**